### ISSUE TITLE
Failure to add interfaces from Mikrotik devices, seemingly random

### ISSUE THREAD

**Issue by ibatten** [Cacti/cacti#3170]
Cacti 1.2.7, but this has been happening for about a year (I just haven't got around to investigating it).

Mikrotik RouterOS 6.46 (ditto)

Adding SNMP-managed Mikrotik equipment sometimes fails to pick up the interfaces, with the SNMP Interface Statistics query reporting Success [0 Items, 0 Rows].  Looking at the results of verbose query, it says:

Total: 0.000000, Delta: 0.000000, XML file parsed ok.
Total: 0.060000, Delta: 0.060000, Executing SNMP get for num of indexes @ '.1.3.6.1.2.1.2.1.0' Index Count: 11
Total: 0.120000, Delta: 0.070000, Executing SNMP walk for list of indexes @ '.1.3.6.1.2.1.2.2.1.1' Index Count: 0
Total: 0.130000, Delta: 0.000000, No SNMP data returned

I have three "identical" (in that they are the same hardware, auto-provisioned by CAPSman with a standard template dropped on top) access points.  Two of them (backroom, upstairs-back) correctly populate the interfaces, the third (frontroom) doesn't.

All of them report the same number of interfaces:

igb@offsite9:$ for i in frontroom backroom upstairs-back; do
snmpbulkwalk -v2c -c XXX udp:$i .1.3.6.1.2.1.2.1.0
done
IF-MIB::ifNumber.0 = INTEGER: 11
IF-MIB::ifNumber.0 = INTEGER: 11
IF-MIB::ifNumber.0 = INTEGER: 11
igb@offsite9:~$ 

but they report them in very different ways:

igb@offsite9:$ for i in frontroom backroom upstairs-back; do
snmpbulkwalk -v2c -c XXX udp:$i .1.3.6.1.2.1.2.2.1.1 | sed "s/^/$i /"
done
frontroom IF-MIB::ifIndex.1 = INTEGER: 1
frontroom IF-MIB::ifIndex.2 = INTEGER: 2
frontroom IF-MIB::ifIndex.3 = INTEGER: 3
frontroom IF-MIB::ifIndex.4 = INTEGER: 4
frontroom IF-MIB::ifIndex.5 = INTEGER: 5
frontroom IF-MIB::ifIndex.40 = INTEGER: 40
frontroom IF-MIB::ifIndex.41 = INTEGER: 41
frontroom IF-MIB::ifIndex.42 = INTEGER: 42
frontroom IF-MIB::ifIndex.43 = INTEGER: 43
frontroom IF-MIB::ifIndex.44 = INTEGER: 44
frontroom IF-MIB::ifIndex.45 = INTEGER: 45
backroom IF-MIB::ifIndex.1 = INTEGER: 1
backroom IF-MIB::ifIndex.2 = INTEGER: 2
backroom IF-MIB::ifIndex.3 = INTEGER: 3
backroom IF-MIB::ifIndex.4 = INTEGER: 4
backroom IF-MIB::ifIndex.5 = INTEGER: 5
backroom IF-MIB::ifIndex.15 = INTEGER: 15
backroom IF-MIB::ifIndex.16 = INTEGER: 16
backroom IF-MIB::ifIndex.17 = INTEGER: 17
backroom IF-MIB::ifIndex.18 = INTEGER: 18
backroom IF-MIB::ifIndex.19 = INTEGER: 19
backroom IF-MIB::ifIndex.20 = INTEGER: 20
upstairs-back IF-MIB::ifIndex.1 = INTEGER: 1
upstairs-back IF-MIB::ifIndex.2 = INTEGER: 2
upstairs-back IF-MIB::ifIndex.3 = INTEGER: 3
upstairs-back IF-MIB::ifIndex.4 = INTEGER: 4
upstairs-back IF-MIB::ifIndex.5 = INTEGER: 5
upstairs-back IF-MIB::ifIndex.2682 = INTEGER: 2682
upstairs-back IF-MIB::ifIndex.2683 = INTEGER: 2683
upstairs-back IF-MIB::ifIndex.2684 = INTEGER: 2684
upstairs-back IF-MIB::ifIndex.2685 = INTEGER: 2685
upstairs-back IF-MIB::ifIndex.2686 = INTEGER: 2686
upstairs-back IF-MIB::ifIndex.2687 = INTEGER: 2687
igb@offsite9:~$ 

Under "Associated Data Queries" for interfaces, backroom and upstairs-back say "Success [82 Items, 9 Rows]" but verbose query says 

Total: 0.050000, Delta: 0.050000, Executing SNMP get for num of indexes @ '.1.3.6.1.2.1.2.1.0' Index Count: 11
Total: 0.150000, Delta: 0.100000, Executing SNMP walk for list of indexes @ '.1.3.6.1.2.1.2.2.1.1' Index Count: 0

and I can create graphs.  frontroom says "Success [0 Items, 0 Rows]", verbose query again says 11 and 0, and I cannot create graphs (as there are no rows).

I now have a similar problem (which I care more about) with a pair of routers: both seem to give similar results when I run the queries by hand, but one of them reports 0 rows in the data query.

Any ideas?  I presume the Mikrotik MIB is doing something unexpected, but I  can't for the life of me see what.

**Comment by ibatten** [comment]
As ever, opening an issue motivates you to investigate more seriously :-)

I replaced snmp{bulkwalk,walk,get,getnext} with scripts which logged their arguments and output.  I found this problem:

Tue Jan  7 12:01:32 GMT 2020 NAME snmpbulkwalk ARGS "-O QnU -c XXX -v 2c -t 1 -r 3 -Cr100 udp:rb450gx4-1.XXX:161 .1.3.6.1.2.1.2.2.1.1" LINES 1
.1.3.6.1.2.1.2.2.1.1 = No Such Object available on this agent at this OID

Running the command by hand shows the issue:

igb@offsite9:share/cacti$ snmpbulkwalk -O QnU -c XXX -v 2c -t 1 -r 3 -Cr100 udp:rb450gx4-1.XXX:161 .1.3.6.1.2.1.2.2.1.1
Error in packet.
Reason: (tooBig) Response message would have been too large.
.1.3.6.1.2.1.2.2.1.1 = No Such Object available on this agent at this OID
igb@offsite9:share/cacti$

Changing the bulkwalk size to -Cr10 fixed it (and for everything else).  A binary chop will presumably find if I can make it slightly larger.
